pkg/serializer/encode.go

```diff
@@ -98,6 +117,117 @@ func (e *encoder) EncodeForGroupVersion(fw FrameWriter, obj runtime.Object, gv s
+	// TODO: Maybe use named returns here?
+
+	// Cast the object to a metav1.Object to get access to annotations
+	metaobj, ok := obj.(metav1.Object)
+	if !ok { // for objects without ObjectMeta, this will fail. allow that failure.
+		// fallback to "normal" encoding
+		return nil, false // fmt.Errorf("couldn't convert to metav1.Object") // TODO: maybe typed errors?
+	}
+
+	// Get annotations and the specific annotation value we're interested in
+	// TODO: This will error if metav1.ObjectMeta is embedded int the object
+	// as a pointer (i.e. *metav1.ObjectMeta) and nil
+	a := metaobj.GetAnnotations()
+	if a == nil {
+		// fallback to "normal" encoding
+		return nil, false // maybe error?
+	}
+	encodedPriorData, ok := a[preserveCommentsAnnotation]
+	if !ok {
+		// fallback to "normal" encoding
+		return nil, false // maybe error?
+	}
+
+	// Decode the base64-encoded bytes of the original object (including the comments)
+	priorData, err := base64.StdEncoding.DecodeString(encodedPriorData)
+	if err != nil {
+		// fatal error
+		return err, true
+	}
+
+	// Unmarshal the original YAML document into a *yaml.RNode, including comments
+	priorNode, err := yaml.Parse(string(priorData))
+	if err != nil {
+		// fatal error
+		return err, true
+	}
+
+	// Make sure we sanitize the annotation before encoding
+	sanitizeAnnotations(obj)
+
+	// Encode the new object into a temporary buffer, it should not be written as the "final result" to the fw
+	buf := new(bytes.Buffer)
+	if err := versionEncoder.Encode(obj, NewYAMLFrameWriter(buf)); err != nil {
+		// fatal error
+		return err, true
+	}
+	updatedData := buf.Bytes()
+
+	// Parse the new, upgraded, encoded YAML into *yaml.RNode for addition
+	// of comments from prevNode
+	afterNode, err := yaml.Parse(string(updatedData))
+	if err != nil {
+		// fatal error
+		return err, true
+	}
+
+	// Copy over comments from the old to the new schema
+	// TODO: Also preserve comments that are "lost on the way", i.e. on schema changes
+	if err := comments.CopyComments(priorNode, afterNode); err != nil {
+		// fatal error
+		return err, true
+	}
+
+	// Print the new schema with the old comments kept to the FrameWriter
+	_, err = fmt.Fprint(fw, afterNode.MustString())
+	// we're done, exit the encode function
+	return err, true
+}
+
+func sanitizeAnnotations(obj runtime.Object) {
+	// TODO: Also automatically sanitize metadata.creationTimestamp which is otherwise always output?
+
+	// Cast the object to a metav1.Object to get access to annotations
+	metaobj, ok := obj.(metav1.Object)
+	if !ok { // if the object doesn't have this ObjectMeta, never mind
+		return
+	}
+
+	// Get annotations and the specific annotation value we're interested in
+	// TODO: This will error if metav1.ObjectMeta is embedded int the object
+	// as a pointer (i.e. *metav1.ObjectMeta) and nil
+	a := metaobj.GetAnnotations()
+	if a == nil { // if the object doesn't have any annotations, never mind
+		return
+	}
+	_, ok = a[preserveCommentsAnnotation]
```

Comment:
This code segment is repeated in quite a lot of places, maybe unify into a private function?